### PR TITLE
Feature/pom parser variable replacement

### PR DESCRIPTION
**Pull request by camperdue42**
Alter the POM parser in order to replace variables within dependencies with the actual desired version.

### CODE REVIEW

File: src/main/java/dev/mccue/resolve/maven/ModelParseException.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package dev.mccue.resolve.maven;
+
+public class ModelParseException extends Throwable{
```
Comment (bowbahdoe) [inline src/main/java/dev/mccue/resolve/maven/ModelParseException.java]: Extend `Exception` if you want a checked exception. `RuntimeException` if not

File: src/main/java/dev/mccue/resolve/maven/PomParser.java [hunk 2]
```diff
@@ -149,7 +150,23 @@ Project project() {
                 versionOpt = Optional.of(parentVersion).filter(s -> !s.isEmpty());
             }
 
-            var properties0 = List.copyOf(properties);
+            var dependencies0 = new ArrayList<Tuple2<Configuration, Dependency>>();
+
+            for (var dependency : dependencies) {
+                final var matcher = Pattern.compile("\\$\\{(.*?)\\}").matcher(dependency.second().version());
```
Comment (bowbahdoe) [inline src/main/java/dev/mccue/resolve/maven/PomParser.java:175]: Can these substitutions be combined with other stuff? like `${some.thing}.5`?

File: src/main/java/dev/mccue/resolve/maven/PomParser.java [hunk 3]
```diff
@@ -149,7 +150,23 @@ Project project() {
...
+                if (matcher.find()) {
+                        final var variable = matcher.group(1);
+                        if (state.properties.containsKey(variable)) {
+                               dependencies0.add(new Tuple2<Configuration, Dependency>(dependency.first(), dependency.second().withVersion(state.properties.get(variable)))); 
+                        } else {
+                                throw new ModelParseException("Undefined variable used in the POM");
```
Comment (bowbahdoe) [inline src/main/java/dev/mccue/resolve/maven/PomParser.java]: Maybe include the name of the variable

File: src/test/java/dev/mccue/resolve/maven/PomParserTest.java [hunk 4]
```diff
@@ -48,7 +51,7 @@ public void parseBasicPOM() throws ParserConfigurationException, SAXException, I
                         <dependency>
```
Comment (bowbahdoe) [inline src/test/java/dev/mccue/resolve/maven/PomParserTest.java]: Instead of changing this test, add another test

**Review by bowbahdoe — CHANGES_REQUESTED** [review]

**Review by zsteig — APPROVED** [review]

**Review by bowbahdoe — APPROVED** [review]

**Pull request merged**